File: Sources/structly/02_hashing/Intersection.swift
//
//  Intersection.swift
//  MyCLI
//
//  Created by new on 3/12/26.
//

func intersection(_ a: [Int], _ b: [Int]) -> [Int] {
    var m = [Int: Int]()
    for x in a {
        m[x, default: 0] += 1
    }
    var res = [Int]()
    for x in b {
        if let fq = m[x] {
            res.append(x)
            m[x] = fq - 1
            if m[x] == 0 {
                m.removeValue(forKey: x)
            }
        }
    }
    
    return res.sorted()
}

File: Sources/structyC/02_hashing/intersection.cpp
//
//  unorderedSet.h
//  MyCLI
//
//  Created by new on 3/12/26.
//

#include <vector>
#include <unordered_set>

std::vector<int> intersection(std::vector<int> a, std::vector<int> b) {
    std::unordered_set<int> set{};
    for (int x : b) {
        set.insert(x);
    }
    std::vector<int> res{};
    for (int x : a) {
        if (set.find(x) != set.end()) {
            res.push_back(x);
        }
    }
    return res;
}
